lib/valueflow.cpp

```diff
@@ -5619,47 +5619,65 @@ static void valueFlowForwardConst(Token* start,
 {
     if (!precedes(start, end))
         throw InternalError(var->nameToken(), "valueFlowForwardConst: start token does not precede the end token.");
-    for (Token* tok = start; tok != end; tok = tok->next()) {
-        if (tok->varId() == var->declarationId()) {
-            for (const ValueFlow::Value& value : values)
-                setTokenValue(tok, value, settings);
-        } else {
-            [&] {
-                // Follow references
-                auto refs = followAllReferences(tok);
-                auto it = std::find_if(refs.cbegin(), refs.cend(), [&](const ReferenceToken& ref) {
-                    return ref.token->varId() == var->declarationId();
-                });
-                if (it != refs.end()) {
-                    for (ValueFlow::Value value : values) {
-                        if (refs.size() > 1)
-                            value.setInconclusive();
-                        value.errorPath.insert(value.errorPath.end(), it->errors.cbegin(), it->errors.cend());
-                        setTokenValue(tok, std::move(value), settings);
+    for (const ValueFlow::Value& value : values) {
+        ProgramMemory pm;
+        pm.setValue(var->nameToken(), value);
+
+        for (Token* tok = start; tok != end; tok = tok->next()) {
+            if (tok->varId() == var->declarationId()) {
+                const Scope* scope = tok->scope();
+                bool setValue = true;
+                while (scope && scope->isLocal()) {
+                    if (scope->type == Scope::ScopeType::eIf) {
+                        MathLib::bigint result{};
+                        bool error{};
+                        execute(scope->bodyStart->linkAt(-1)->astOperand2(), pm, &result, &error, settings);
+                        if (!result && !error) {
+                            setValue = false;
+                            break;
+                        }
```

Comment:
This isn't very efficient as you need to evaluate the condition every time there is a matching variable used. It would be better to traverse how `findTokensSkipDeadCode` does, by skipping the branch when it sees an `if` token that evaluates to true or false.

Reply:
If earlier values are replaced by later ones (assuming that they are more correct), shouldn't `valueFlowForwardConst()` just run earlier?